### PR TITLE
Adding Gutter icon in SAM template.

### PR DESCRIPTION
**Pull request by zhangzhx**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Show Run/Debug Gutter in SAM template

## Related Issue(s)

#299

## Screenshots (if appropriate)

screen shot 2018-10-25 at 11 17 27 am

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -84,6 +84,7 @@
         <programRunner implementation="software.aws.toolkits.jetbrains.services.lambda.execution.remote.RemoteLambdaRunner"/>
         <fileBasedIndex implementation="software.aws.toolkits.jetbrains.services.lambda.LambdaHandlerIndex"/>
         <codeInsight.lineMarkerProvider language="" implementationClass="software.aws.toolkits.jetbrains.services.lambda.upload.LambdaLineMarker"/>
+        <runLineMarkerContributor language="" implementationClass="software.aws.toolkits.jetbrains.services.lambda.execution.template.LambdaRunLineMarkerContributor"/>
```
Comment (abrooksv) [inline jetbrains-core/resources/META-INF/plugin.xml]: Language should be YAML

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    override fun getInfo(element: PsiElement): Info? {
+        // Only leaf element is allowed
+        if (element.firstChild != null) {
+            return null
+        }
+
+        val parent = element.parent ?: return null
+        return try {
+            functionFromElement(parent) ?: return null
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt]: Do the expensive checks after all the other cheaper checks are done

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                    if (parent.key != element) {
+                        return null
+                    }
+                }
+                else -> return null
+            }
+
+            Info(AllIcons.RunConfigurations.TestState.Run,
+                    ExecutorAction.getActions(1), null)
+        } catch (e: Exception) {
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt]: What can throw?

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributorTest.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda.execution.template
+
+import assertk.assert
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributorTest.kt]: We are not using assertk anymore, AssertJ from now on

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributorTest.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private fun runAndAssertionMarks(fixture: CodeInsightTestFixture, assertion: (List<LineMarkerInfo.LineMarkerGutterIconRenderer<PsiElement>>) -> Unit) {
+        runInEdtAndWait {
+            val marks = fixture.findAllGutters().filterIsInstance<LineMarkerInfo.LineMarkerGutterIconRenderer<PsiElement>>()
+            assertion(marks)
+        }
+    }
+
+    private fun CodeInsightTestFixture.openFile(relativePath: String, fileText: String): VirtualFile {
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributorTest.kt]: Move this with the other CodeInsightTestFixture extensions in JavaCodeInsightTestFixtureRule

File: jetbrains-core/resources/META-INF/plugin.xml (hunk 1 above)
Comment (zhangzhx) [reply jetbrains-core/resources/META-INF/plugin.xml]: I'd like to use this class to support JSON as well later.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt (hunk 3 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/template/LambdaRunLineMarkerContributor.kt]: UnsupportedOperationException from `functionFromElement`

File: jetbrains-core/resources/META-INF/plugin.xml (hunk 1 above)
Comment (kiiadi) [reply jetbrains-core/resources/META-INF/plugin.xml]: we can register it with both languages when we get there

**Comment by kiiadi** [comment]
awesome!!

**Pull request closed without merging**